zend_abstract_interface/config/php7-8/config_runtime.c

```diff
@@ -10,44 +10,110 @@
 
 extern HashTable zai_config_name_map;
 
-ZAI_TLS zval *runtime_config;  // dynamically allocated, otherwise TLS alignment limits may be exceeded
-ZAI_TLS bool runtime_config_initialized = false;
+ZAI_TLS zval *runtime_config = NULL;  // dynamically allocated, otherwise TLS alignment limits may be exceeded
+
+static void zai_config_runtime_config_update(void);
+
+static inline void zai_config_runtime_dtor(zval *zv) {
+    if (!Z_OPT_REFCOUNTED_P(zv)) {
+        return;
+    }
+
+    if (Z_TYPE_P(zv) == IS_ARRAY) {
+#ifdef IS_ARRAY_PERSISTENT
+        if (!(GC_FLAGS(Z_COUNTED_P(zv)) & IS_ARRAY_PERSISTENT)) {
+#else
+        if (!(Z_ARRVAL_P(zv)->u.flags & HASH_FLAG_PERSISTENT)) {
+#endif
+            zval_ptr_dtor(zv);
+        }
+        return;
+    }
+
+    if (Z_TYPE_P(zv) == IS_STRING) {
+        if (!(GC_FLAGS(Z_COUNTED_P(zv)) & IS_STR_PERSISTENT)) {
+            zval_ptr_dtor(zv);
+        }
+    }
+}
 
 void zai_config_replace_runtime_config(zai_config_id id, zval *value) {
     zval *rt_value = &runtime_config[id];
-    zval_ptr_dtor(rt_value);
+    zai_config_runtime_dtor(rt_value);
```

Comment:
This doesn't sound very good if this needs specialized destruction. Especially if the value is leaked within dd_trace_env_config() into userland and then replaced thanks to an ini set, userland will have that zval and WILL call zval_ptr_dtor on it.